streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -114,18 +114,18 @@ public StreamTask(TaskId id,
         // to corresponding source nodes in the processor topology
         partitionQueues = new HashMap<>();
 
-        TimestampExtractor timestampExtractor = config.getConfiguredInstance(StreamsConfig.TIMESTAMP_EXTRACTOR_CLASS_CONFIG, TimestampExtractor.class);
+        //TimestampExtractor timestampExtractor = config.getConfiguredInstance(StreamsConfig.TIMESTAMP_EXTRACTOR_CLASS_CONFIG, TimestampExtractor.class);
 
         for (TopicPartition partition : partitions) {
             SourceNode source = topology.source(partition.topic());
-            RecordQueue queue = createRecordQueue(partition, source, timestampExtractor);
+            RecordQueue queue = createRecordQueue(partition, source, source.getTimestampExtractor());
```

Comment:
add line:
`TimestampExtractor sourceTimestampExtractor = source.getTimestampExtractor();`
and change to
`RecordQueue queue = createRecordQueue(partition, source, sourceTimestampExtractor != null ? sourceTimestampExtractor : defaultTimestampExtractor);